### PR TITLE
Fixed exceptional case handling in SherpaHadronizer

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

The MPI::Init function setups up a unix signal handler which interferes with the signal handled used in the cmsExternalGenerator program. Moving the call to the constructor allows the signal handler to be reset
before doing any data processing.

Even with this change, an unix signal happening in the constructor will cause a dealock of the ExternalGenerator system.

Also had to avoid a case where an exception could be thrown while we were unrolling from another exception. That would happen if doSetRandomEngine would throw an exception.

#### PR validation:

The code compiles and a simple job using the SherpaHadronizer with the ExternalGeneratorFilter ran fine.
Explicitly adding an exception to Sherpa allowed us to test the problem.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
type bugfix

**Comment by Dr15Jones** [comment]
urgent

we will need a backport to 10_6 for production.

**Comment by GurpreetSinghChahal** [comment]
+1

**Comment by perrotta** [comment]
+1

File: GeneratorInterface/SherpaInterface/src/SherpaHadronizer.cc [hunk 1]
```diff
@@ -210,14 +218,12 @@ SherpaHadronizer::~SherpaHadronizer() {
 
 bool SherpaHadronizer::initializeForInternalPartons() {
   //initialize Sherpa but only once
+  throw cms::Exception("TEST");
```
Comment (Dr15Jones) [inline GeneratorInterface/SherpaInterface/src/SherpaHadronizer.cc:221]: Argg! This was left over from my testing.

Comment (Dr15Jones) [reply GeneratorInterface/SherpaInterface/src/SherpaHadronizer.cc:221]: Fixed in #35224

**Pull request merged**